superset-frontend/src/explore/reducers/exploreReducer.js

```diff
@@ -99,24 +99,34 @@ export default function exploreReducer(state = {}, action) {
     },
     [actions.SET_FIELD_VALUE]() {
       const new_form_data = state.form_data;
-      new_form_data[action.controlName] = action.value;
+      const { controlName, value, validationErrors } = action;
+      new_form_data[controlName] = value;
 
-      // These errors are reported from the Control components
-      let errors = action.validationErrors || [];
       const vizType = new_form_data.viz_type;
+
       // Use the processed control config (with overrides and everything)
       // if `controlName` does not existing in current controls,
       const controlConfig =
         state.controls[action.controlName] ||
         getControlConfig(action.controlName, vizType) ||
         {};
+
+      // will call validators again
       const control = {
         ...getControlStateFromControlConfig(controlConfig, state, action.value),
       };
 
-      // These errors are based on control config `validators`
-      errors = errors.concat(control.validationErrors || []);
+      // combine newly detected errors with errors from `onChange` event of
+      // each control component (passed via reducer action).
```

Comment:
Sometimes the `onChange` event returns the same validation error than the one returned by validators. (**Maybe we shouldn't allow controls to run validations in `onChange`.**)

Reply:
> Sometimes the onChange event returns the same validation error than the one returned by validators.

The same one or different one?

Reply:
I think most cases the same. I only checked `isInteger` validator for calendar chart.